gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java

```diff
@@ -73,17 +116,104 @@ public void close() {
       if (instance == null) {
         throw new IllegalStateException("GcpObservability already closed!");
       }
+      unRegisterStackDriverExporter();
       LoggingChannelProvider.shutdown();
       LoggingServerProvider.shutdown();
       sink.close();
       instance = null;
     }
   }
 
-  private GcpObservability(Sink sink,
+  private void setProducer(
+      InternalLoggingChannelInterceptor.Factory channelInterceptorFactory,
+      InternalLoggingServerInterceptor.Factory serverInterceptorFactory) {
+    if (config.isEnableCloudLogging()) {
+      clientInterceptors.add(channelInterceptorFactory.create());
+      serverInterceptors.add(serverInterceptorFactory.create());
+    }
+    if (config.isEnableCloudMonitoring()) {
+      clientInterceptors.add(
+          InternalCensusStatsAccessor.getClientInterceptor(true, true, true, true));
+      tracerFactories.add(
+          InternalCensusStatsAccessor.getServerStreamTracerFactory(true, true, true));
+    }
+    if (config.isEnableCloudTracing()) {
+      clientInterceptors.add(InternalCensusTracingAccessor.getClientInterceptor());
+      tracerFactories.add(InternalCensusTracingAccessor.getServerStreamTracerFactory());
+    }
+
+    if (!clientInterceptors.isEmpty()
+        || !serverInterceptors.isEmpty()
+        || !tracerFactories.isEmpty()) {
+      InternalGlobalInterceptors.setInterceptorsTracers(
+          clientInterceptors, serverInterceptors, tracerFactories);
+    }
+  }
+
+  private void registerStackDriverExporter(String projectId) throws IOException {
+    if (config.isEnableCloudMonitoring()) {
+      RpcViews.registerAllGrpcViews();
+      StackdriverStatsConfiguration.Builder statsConfigurationBuilder =
+          StackdriverStatsConfiguration.builder();
+      if (!Strings.isNullOrEmpty(projectId)) {
+        statsConfigurationBuilder.setProjectId(projectId);
+      }
+      try {
+        StackdriverStatsExporter.createAndRegister(statsConfigurationBuilder.build());
+      } catch (IOException e) {
+        throw new IOException("Failed to register Stackdriver stats exporter, " + e.getMessage());
```

Comment:
This is unusual: catching an `IOException` and throwing another `IOException` with just the message modified. Wouldn't the original message be sufficient - so no need to do this?

Reply:
Also I realized about the `IllegalStateException`: so we will get that for duplicate registration attempt. Ideally that should not happen because of the guard around `grpcInit` but a user could have registered the exporter on their own. I guess the only way to resolve that is to document that in the UG?

Reply:
Removed catching and re-throwing exceptions.